File: src/ladym/operations/proceduralize.py
"""L1 → L3 proceduralization (ARCHITECTURE.md §2 ``proceduralize``).

Clusters successful episodic events by similarity and, when a cluster crosses a support
threshold, writes an L3 playbook that captures the recurring procedure. CoALA's skill
library; Voyager's auto-skill authoring.
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field

from ..config import Config
from ..layers.procedural import ProceduralMemory, _playbook_content
from ..layers.semantic import content_hash
from ..schema import Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore
from .consolidate import Action
from .supersedes import retire as _retire


@dataclass
class ProceduralizeReport:
    clusters_examined: int = 0
    playbooks_created: int = 0  # = ADD count (kept for backward compat)
    actions: dict[str, int] = field(
        default_factory=lambda: {"ADD": 0, "UPDATE": 0, "NOOP": 0}
    )
    details: list[dict] = None  # type: ignore[assignment]

    def __post_init__(self):
        if self.details is None:
            self.details = []


def _retrieve_existing_playbooks(
    store: SQLiteStore, candidate_vec: list[float], ws: str, top_k: int
) -> list[tuple[Memory, float]]:
    """Top similar existing L3 playbooks in this workspace (mirrors consolidate's L2 retrieval)."""
    raw = store.vector_index.search(candidate_vec, top_k=top_k)
    similar: list[tuple[Memory, float]] = []
    for mid, sim in raw:
        if sim < 0.1:
            continue
        m = store.get_memory(mid)
        if m is None or m.workspace != ws:
            continue
        if m.layer != Layer.PROCEDURAL.value or m.type != MemoryType.PLAYBOOK.value:
            continue
        similar.append((m, sim))
    similar.sort(key=lambda t: t[1], reverse=True)
    return similar


def _classify_playbook(
    candidate_hash: str,
    similar: list[tuple[Memory, float]],
    threshold: float,
) -> Action:
    """ADD/UPDATE/NOOP for a candidate playbook vs existing L3."""
    for existing, _sim in similar:
        if existing.content_hash and existing.content_hash == candidate_hash:
            return Action.NOOP
    if similar and similar[0][1] >= threshold:
        return Action.UPDATE
    return Action.ADD


def proceduralize(
    store: SQLiteStore,
    embedder: EmbeddingProvider,
    *,
    cfg: Config,
    workspace: str | None = None,
    min_cluster_size: int = 3,
    similarity_threshold: float = 0.55,
) -> ProceduralizeReport:
    ws = workspace or cfg.workspace
    proc = ProceduralMemory(store, embedder, workspace=ws)
    report = ProceduralizeReport()

    rows = store.conn.execute(
        "SELECT * FROM memories WHERE layer = ? AND workspace = ? ORDER BY created_at",
        (Layer.EPISODIC.value, ws),
    ).fetchall()
    episodes = [store._row_to_memory(r) for r in rows]
    # only successful episodes ("outcome=success") carry procedural signal
    succ = [e for e in episodes if e.metadata.get("outcome") in ("success", "ok", "done")]
    if len(succ) < min_cluster_size:
        return report

    vecs = embedder.embed_batch([e.content for e in succ])
    assigned: list[bool] = [False] * len(succ)

    for i, anchor in enumerate(succ):
        if assigned[i]:
            continue
        cluster = [anchor]
        for j in range(i + 1, len(succ)):
            if assigned[j]:
                continue
            from ..storage.embeddings import cosine_similarity
            if cosine_similarity(vecs[i], vecs[j]) >= similarity_threshold:
                cluster.append(succ[j])
                assigned[j] = True
        if len(cluster) >= min_cluster_size:
            assigned[i] = True
            report.clusters_examined += 1
            actions = [c.metadata.get("action", "do") for c in cluster]
            top_action = Counter(actions).most_common(1)[0][0]
            steps = _derive_steps(cluster)
            name = f"How to {top_action} ({len(cluster)} episodes)"
            # idempotency: check existing L3 playbooks before writing
            candidate_content = _playbook_content(name, steps)
            candidate_hash = content_hash(candidate_content)
            candidate_vec = embedder.embed(candidate_content)
            similar = _retrieve_existing_playbooks(
                store, candidate_vec, ws,
                top_k=cfg.consolidate.min_episodes_to_trigger + 5,
            )
            action = _classify_playbook(candidate_hash, similar, similarity_threshold)
            report.actions[action.value] += 1
            if action == Action.ADD:
                proc.put_playbook(
                    name=name, steps=steps,
                    preconditions=list({c.metadata.get("agent", "agent") for c in cluster}),
                    expected_outcome="success",
                    tags=[top_action],
                )
                report.playbooks_created += 1
            elif action == Action.UPDATE and similar:
                new_mem = proc.put_playbook(
                    name=name, steps=steps,
                    preconditions=list({c.metadata.get("agent", "agent") for c in cluster}),
                    expected_outcome="success",
                    tags=[top_action],
                )
                _retire(store, similar[0][0], new_id=new_mem.id)
            # NOOP: skip
            report.details.append({"action": action.value, "action_verb": top_action, "size": len(cluster)})
    return report


def _derive_steps(cluster: list) -> list[str]:
    """Heuristic: turn the distinct (action, observation) pairs of a cluster into ordered steps."""
    seen: set[str] = set()
    steps: list[str] = []
    for c in cluster:
        key = (c.metadata.get("action", ""), c.metadata.get("observation", ""))
        s = f"{key[0]}" + (f" — {key[1]}" if key[1] else "")
        if s and s not in seen:
            seen.add(s)
            steps.append(s)
    return steps

File: operations/proceduralize.go
package operations

import (
	"sort"
	"strconv"

	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/layers"
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// ProceduralizeReport reports the outcome of a proceduralization pass.
type ProceduralizeReport struct {
	ClustersExamined int
	PlaybooksCreated int
	Actions          map[string]int
	Details          []map[string]any
}

func newProceduralizeReport() *ProceduralizeReport {
	return &ProceduralizeReport{Actions: map[string]int{
		string(ActionAdd): 0, string(ActionUpdate): 0, string(ActionNoop): 0,
	}}
}

func retrieveExistingPlaybooks(store storage.Store, candidateVec []float32, ws string, topK int) ([]similarFact, error) {
	raw := store.VectorSearch(candidateVec, topK)
	var similar []similarFact
	for _, h := range raw {
		if h.Similarity < 0.1 {
			continue
		}
		m, err := store.GetMemory(h.ID)
		if err != nil {
			return nil, err
		}
		if m == nil || m.Workspace != ws {
			continue
		}
		if m.Layer != schema.LayerProcedural || m.Type != schema.TypePlaybook {
			continue
		}
		if IsRetired(m) {
			continue
		}
		similar = append(similar, similarFact{Memory: m, Sim: h.Similarity})
	}
	sort.SliceStable(similar, func(a, b int) bool { return similar[a].Sim > similar[b].Sim })
	return similar, nil
}

func classifyPlaybook(candidateHash string, similar []similarFact, threshold float64) Action {
	for _, s := range similar {
		if s.Memory.ContentHash != "" && s.Memory.ContentHash == candidateHash {
			return ActionNoop
		}
	}
	if len(similar) > 0 && similar[0].Sim >= threshold {
		return ActionUpdate
	}
	return ActionAdd
}

// Proceduralize clusters successful episodic events into L3 playbooks.
func Proceduralize(store storage.Store, embedder storage.EmbeddingProvider, cfg *config.Config, workspace string, minClusterSize int, similarityThreshold float64) (*ProceduralizeReport, error) {
	ws := workspace
	if ws == "" {
		ws = cfg.Workspace
	}
	if minClusterSize == 0 {
		minClusterSize = 3
	}
	if similarityThreshold == 0 {
		similarityThreshold = 0.55
	}
	proc := layers.NewProceduralMemory(store, embedder, ws)
	report := newProceduralizeReport()

	episodes, err := store.IterMemories(ws, string(schema.LayerEpisodic), "")
	if err != nil {
		return nil, err
	}
	sort.SliceStable(episodes, func(a, b int) bool { return episodes[a].CreatedAt < episodes[b].CreatedAt })

	var succ []*schema.Memory
	for _, e := range episodes {
		outcome := e.MetaString("outcome")
		if outcome == "success" || outcome == "ok" || outcome == "done" {
			succ = append(succ, e)
		}
	}
	if len(succ) < minClusterSize {
		return report, nil
	}

	contents := make([]string, len(succ))
	for i, e := range succ {
		contents[i] = e.Content
	}
	vecs, err := embedder.EmbedBatch(contents)
	if err != nil {
		return nil, err
	}
	assigned := make([]bool, len(succ))

	for i, anchor := range succ {
		if assigned[i] {
			continue
		}
		cluster := []*schema.Memory{anchor}
		for j := i + 1; j < len(succ); j++ {
			if assigned[j] {
				continue
			}
			if storage.CosineSimilarity(vecs[i], vecs[j]) >= similarityThreshold {
				cluster = append(cluster, succ[j])
				assigned[j] = true
			}
		}
		if len(cluster) >= minClusterSize {
			assigned[i] = true
			report.ClustersExamined++
			// Python: Counter(c.metadata.get("action", "do") ...).most_common(1)
			// — missing action counts as "do", ties keep first-occurrence order.
			actionCounts := map[string]int{}
			var actionOrder []string
			for _, c := range cluster {
				a := c.MetaString("action")
				if a == "" {
					a = "do"
				}
				if _, ok := actionCounts[a]; !ok {
					actionOrder = append(actionOrder, a)
				}
				actionCounts[a]++
			}
			topAction := ""
			topN := 0
			for _, a := range actionOrder {
				if actionCounts[a] > topN {
					topAction = a
					topN = actionCounts[a]
				}
			}
			steps := deriveSteps(cluster)
			name := "How to " + topAction + " (" + strconv.Itoa(len(cluster)) + " episodes)"
			candidateContent := layers.PlaybookContent(name, steps)
			candidateHash := schema.ContentHash(candidateContent)
			candidateVec, err := embedder.Embed(candidateContent)
			if err != nil {
				return nil, err
			}
			similar, err := retrieveExistingPlaybooks(store, candidateVec, ws, cfg.Consolidate.MinEpisodesToTrigger+5)
			if err != nil {
				return nil, err
			}
			action := classifyPlaybook(candidateHash, similar, similarityThreshold)
			report.Actions[string(action)]++

			preconditions := []string{}
			seenAgents := map[string]bool{}
			for _, c := range cluster {
				a := c.MetaString("agent")
				if a == "" {
					a = "agent"
				}
				if !seenAgents[a] {
					seenAgents[a] = true
					preconditions = append(preconditions, a)
				}
			}

			if action == ActionAdd {
				if _, err := proc.PutPlaybook(name, steps, preconditions, "success", []string{topAction}); err != nil {
					return nil, err
				}
				report.PlaybooksCreated++
			} else if action == ActionUpdate && len(similar) > 0 {
				newMem, err := proc.PutPlaybook(name, steps, preconditions, "success", []string{topAction})
				if err != nil {
					return nil, err
				}
				if err := Retire(store, similar[0].Memory, newMem.ID); err != nil {
					return nil, err
				}
			}
			report.Details = append(report.Details, map[string]any{
				"action": string(action), "action_verb": topAction, "size": len(cluster),
			})
		}
	}
	return report, nil
}

func deriveSteps(cluster []*schema.Memory) []string {
	seen := map[string]bool{}
	var steps []string
	for _, c := range cluster {
		action := c.MetaString("action")
		observation := c.MetaString("observation")
		s := action
		if observation != "" {
			s += " — " + observation
		}
		if s != "" && !seen[s] {
			seen[s] = true
			steps = append(steps, s)
		}
	}
	return steps
}
